### PR TITLE
Add FastText classifier as a new component

### PR DESCRIPTION
**Pull request by yfletberliac**
**Proposed changes**:
- add fasttext wrapper as a classifier based on https://github.com/facebookresearch/fastText/blob/master/python/fastText/FastText.py

**Status (please check what you already did)**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog

### CODE REVIEW

**Comment by tmbo** [comment]
cool stuff! one question: This doesn't train the classifier though, right? So you'd need to train that separately or how does it work?

**Comment by yfletberliac** [comment]
Hi @tmbo, thanks. Indeed, this doesn't train the classifier, you would need to train it beforehand or use a pre-trained model.
The resulting _model.ftz_ (quantized) or _model.bin_ file should be put under projects/default and the inference will work normally as it is working with any other classifier (e.g. sklearn).

**Comment by yfletberliac** [comment]
Hi @tmbo, could I have an update on this PR?
Best,
Yannis

**Comment by yfletberliac** [comment]
Hi @tmbo,
In fact the component **FastTextIntentClassifier** created here is a a component for a [**fasttext classification model**](https://fasttext.cc/docs/en/supervised-tutorial.html).
There are two types of application with FastText:
- word vectors, which I agree don't need to be covered by a new component since they can be accessible through a spacy model);
- a **classification model**, which I implemented the corresponding component here.

The classification model needs to be trained with labeled data beforehand (resulting in a model.ftz file), which is directly accessible and usable with the **FastTextIntentClassifier** component.

Nevertheless, I agree that the training part is not implemented yet.

**Comment by Ghostvv** [comment]
I think it makes sense to add fasttext wrapper as a classifier, but the train method should be implemented.
According to https://github.com/facebookresearch/fastText/blob/master/python/fastText/FastText.py, fastext has its own tokenizer, is it bad? Why would you need nltk tokenizer?

File: rasa_nlu/classifiers/fasttext_intent_classifier.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    name = "intent_classifier_fasttext"
+
+    provides = ["intent", "intent_ranking"]
+
+    def __init__(self, clf=None, le=None):
+        # type: (fastText) -> None
+        """Construct a new intent classifier using the fasttext framework."""
+
+        self.clf = clf
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/fasttext_intent_classifier.py]: Are there any configuration parameters, for example language shall be passed through config

File: rasa_nlu/classifiers/fasttext_intent_classifier.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    @classmethod
+    def required_packages(cls):
+        # type: () -> List[Text]
+        return ["numpy", "fastText"]
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/fasttext_intent_classifier.py]: numpy should not be here

File: rasa_nlu/classifiers/fasttext_intent_classifier.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        :param labels: List of labels to convert to numeric representation"""
+
+        return [x.replace('__label__', '') for x in labels]
+
+    def train(self, training_data, config, **kwargs):
+        # type: (TrainingData, RasaNLUConfig, **Any) -> None
+        """Train the intent classifier on a data set.
+        :param num_threads: number of threads used during training time
+        We keep it here for later use."""
+
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/fasttext_intent_classifier.py]: please call fasttext train method here

File: rasa_nlu/classifiers/fasttext_intent_classifier.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    def persist(self, model_dir):
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/fasttext_intent_classifier.py]: please add ability to save fasttext model

File: rasa_nlu/classifiers/fasttext_intent_classifier.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            word_list = nltk.word_tokenize(raw_text.decode('utf-8'), 'english')  # Tokenize with nltk english
+        elif language=="fr_FR":
+            word_list = nltk.word_tokenize(raw_text.decode('utf-8'), 'french')  # Tokenize with nltk french
+        else:
+            raise ValueError('nltk.word_tokenize: language in model_metadata not covered.')
+
+        if language=="en_EN":
+            stopword_set = set(stopwords.words("english"))  # List of stopwords in english
+        elif language=="fr_FR":
+            stopword_set = set(stopwords.words("french"))  # List of stopwords in french
```
Comment (Ghostvv) [inline rasa_nlu/classifiers/fasttext_intent_classifier.py]: it would be good to support all languages supported by nltk stopwords

**Comment by Ghostvv** [comment]
@yfletberliac also did you try `tensor flow_embedding` pipeline for your classification task?

**Comment by amn41** [comment]
thanks @yfletberliac for taking the time to work on this. 

As of now, I don't think we will want to integrate and maintain this. Adding a new component means that we take on a lot of future work to test and maintain it, so it needs to provide a big benefit. I suspect that the new `tensorflow_embedding` pipeline will solve many of the things this aims to do. However, if there are major things still lacking, or significant performance differences, I'm open to looking into this more.

**Comment by yfletberliac** [comment]
Hi @Ghostvv,

> According to https://github.com/facebookresearch/fastText/blob/master/python/fastText/FastText.py, fastext has its own tokenizer, is it bad? Why would you need nltk tokenizer?

FastText tokenizer is very simple: it tokenizes only on whitespaces. The nltk tokenizer addresses this issue (see https://github.com/facebookresearch/fastText/issues/175).

> @yfletberliac also did you try tensorflow_embedding pipeline for your classification task?

No, I haven't tried it. Since I'm already using pretrained embedding vectors, I'm not sure it would give better results that I have... but it could worth the try.

**Comment by Ghostvv** [comment]
@yfletberliac thanks for explaining tokenization issue. If you have time, I'd be really curious how performance of `tensorflow_embedding` pipeline compares to your results

**Comment by amn41** [comment]
as I said, @yfletberliac - I will close this for now. Please create an issue if you see significant performance improvements.

**Pull request closed without merging**